wagtail/images/models.py

```diff
@@ -300,93 +309,166 @@ def get_rendition_model(cls):
         """Get the Rendition model for this Image model"""
         return cls.renditions.rel.related_model
 
-    def get_rendition(self, filter):
+    def get_rendition(self, filter: Union["Filter", str]) -> "AbstractRendition":
+        """
+        Returns a ``Rendition``* instance with a ``file`` field value (an
+        image) reflecting the supplied ``filter`` value and focal point values
+        from this object.
+
+        *If using custom image models, an instance of the custom rendition
+        model will be returned.
+        """
         if isinstance(filter, str):
             filter = Filter(spec=filter)
 
-        cache_key = filter.get_cache_key(self)
         Rendition = self.get_rendition_model()
 
         try:
-            rendition_caching = True
+            rendition = self.find_existing_rendition(filter)
+        except Rendition.DoesNotExist:
+            rendition = self.create_rendition(filter)
+            # Reuse this rendition if requested again from this object
+            if "renditions" in getattr(self, "_prefetched_objects_cache", {}):
+                self._prefetched_objects_cache["renditions"]._result_cache.append(
+                    rendition
+                )
+
+        try:
             cache = caches["renditions"]
-            rendition_cache_key = Rendition.construct_cache_key(
-                self.id, cache_key, filter.spec
+            key = Rendition.construct_cache_key(
+                self.id, filter.get_cache_key(self), filter.spec
             )
-            cached_rendition = cache.get(rendition_cache_key)
+            cache.set(key, rendition)
+        except InvalidCacheBackendError:
+            pass
+
+        return rendition
+
+    def find_existing_rendition(self, filter: "Filter") -> "AbstractRendition":
+        """
+        Returns an existing ``Rendition``* instance with a ``file`` field value
+        (an image) reflecting the supplied ``filter`` value and focal point
+        values from this object.
+
+        If no such rendition exists, a ``DoesNotExist`` error is raised for the
+        relevant model.
+
+        *If using custom image models, an instance of the custom rendition
+        model will be returned.
+        """
+
+        Rendition = self.get_rendition_model()
+        cache_key = filter.get_cache_key(self)
+
+        # Interrogate prefetched values first (if available)
+        if "renditions" in getattr(self, "_prefetched_objects_cache", {}):
+            for rendition in self.renditions.all():
```

Comment:
question: do we need to iterate through all the renditions?
would using something like `self.renditions.filter(filter_spec=filter.spec, focal_point_key=cache_key).first()` not do?

Reply:
Yes, applying `filter()` to any `QuerySet` returns a completely new one which, if evaluated, would result in another database query. Likewise, `first()` always results in a new database query too.

Reply:
Heh, didn't realise I posted a full comment.
Was adding draft questions as I went. You are very right in this. We're dealing with prefetched values, so they'd naturally be cached. 🌟